core/src/main/scala/org/apache/spark/sql/internal/session.scala

```diff
@@ -515,6 +515,21 @@ private[sql] final class PreprocessTable(state: SnappySessionState) extends Rule
     case c@CreateTable(tableDesc, mode, queryOpt) if DDLUtils.isDatasourceTable(tableDesc) =>
       val tableIdent = state.catalog.resolveTableIdentifier(tableDesc.identifier)
       val provider = tableDesc.provider.get
+      if (mode == SaveMode.Overwrite && queryOpt.isDefined && queryOpt.get.resolved) {
+        EliminateSubqueryAliases(queryOpt.get) match {
+          case plan: LogicalPlan =>
+            val srcRelationNames = plan.collect {
+              case LogicalRelation(src: MutableRelation, _, _) => src.table
+            }
+            // we store table as Uppercase in relation
+            if (srcRelationNames.contains(tableIdent.unquotedString.toUpperCase)) {
+              throw new AnalysisException(
+                s"Cannot overwrite table ${tableDesc.identifier} that is also being read from")
+            }
```

Comment:
Why only check for MutableRelation because the behaviour seems to be a problem for other data sources too. Since PreWriteCheck is already applied that does the catalog lookup, I think the safest option is to add a simple SessionCatalogWrapper around SnappySessionCatalog that overrides lookupRelation to instead invoke resolveRelationWithAlias. This wrapper catalog can be the one passed to PreWriteCheck constructor in SnappySessionState/SnappyAQPSessionState.getExtendedCheckRules.